ent/encryption/encrypted_file_impl.cc

```diff
@@ -701,5 +701,82 @@ std::unique_ptr<data_sink_impl> make_encrypted_sink(data_sink sink, shared_ptr<s
     return std::make_unique<encrypted_data_sink>(std::move(sink), std::move(k));
 }
 
+class encrypted_data_source : public data_source_impl, public block_encryption_base {
+    data_source _source;
+    std::vector<char> _buffer;
+    size_t current_position = 0;
+
+public:
+    encrypted_data_source(data_source source, shared_ptr<symmetric_key> k) : block_encryption_base(std::move(k)), _source(std::move(source)) {
+        _buffer.reserve(block_size * 2);
+    }
+
+    future<temporary_buffer<char>> get() override {
+        while (_buffer.size() < block_size) {
+            auto buf = co_await _source.get();
+            if (buf.empty()) {
+                break;
+            }
+            std::move(std::make_move_iterator(buf.get_write()), std::make_move_iterator(buf.get_write() + buf.size()), std::back_inserter(_buffer));
+        }
+
+        const size_t key_block_size = _key->block_size();
+        const size_t bytes_to_consume = _buffer.size() < block_size ? align_down(_buffer.size(), key_block_size) : align_down(_buffer.size(), block_size);
+        temporary_buffer<char> output(bytes_to_consume);
+        size_t decrypted_bytes = 0;
+
+        for (size_t offset = 0; offset < bytes_to_consume; offset += block_size) {
+            auto iv = iv_for(current_position + offset);
+            // Determine how many bytes remain in this block
+            const size_t remaining = std::min<uint64_t>(block_size, bytes_to_consume - offset);
+
+            // If the remaining bytes form a partial block, align down to a whole block size
+            size_t bytes_to_decrypt = (remaining < block_size) ? align_down(remaining, key_block_size) : block_size;
+
+            _key->transform_unpadded(mode::decrypt, _buffer.data() + offset, bytes_to_decrypt, output.get_write() + offset, iv.data());
+            decrypted_bytes += bytes_to_decrypt;
```

Comment:
Not quite. If we get here because of EOS, and remaining is not aligned on key block size, the actual data length is `remaining - key_block_size`. See file/sink on how unaligned file sizes are handled.